File: 20_valid_parentheses/stack_brackets.py
import numpy as np
import time
import copy

bracket_map = {'}':'{', ')':'(', ']':'['}

def find_valid_brackets(s):
    stk = []
    for c in s:
        if c not in(bracket_map):
            stk.append(c)
            continue
            
        elif len(stk) == 0:
            return False
        elif bracket_map[c] != stk[-1]:
            return False
        stk.pop()

    if len(stk) == 0:
        return True

    return False

if __name__ == '__main__':
    input_str = '{[]()}'
    print(find_valid_brackets(input_str))
    print("main")

File: 20_valid_parentheses/stack_brackets.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
#include <map>
#include <string>
#include <stack>

using namespace std;

bool is_valid_parentheses(string s)
{
  unordered_map<char, char> brackets {{'{','}'},{'[',']'},{'(',')'}};
  stack<char> stk;
  for(auto it : s)
  {
    if(brackets.find(it) != brackets.end())
    {
      stk.push(it);
      continue;
    }
    else if(stk.empty())
      return false;
    if(brackets[stk.top()] != it)
      return false;

    stk.pop();
  }
  if(stk.empty())
    return true;

  return false;
}

int main()
{
  string input_str = "{()}()";

  if(is_valid_parentheses(input_str))
    cout << "The input string: " << input_str << " is valid!" << endl;

  else
    cout << "The input string: " << input_str << " is not valid." << endl;
  // for(auto it = input_str.begin(); it != input_str.end(); ++it)
  
  return 0;
}